Notebook for Learning Rate Scheduler

ParagEkbote opened the pull request:
As described in #275, I've created a notebook that shows the visualization of the learning rate schedulers using the custom scheduler callback method in skorch.

Please let me know if any improvements are needed and I will make the necessary changes.

Fixes #275 

cc: @BenjaminBossan

BenjaminBossan reviewed (changes requested):
Thanks a lot for the PR. I haven't done an in-depth review yet but the first impression is already quite good.

Before proceeding, I have a few points, could you please take a look?

1. Right now, the notebook starts with `pip install ...`. We typically have a check if the notebook runs in a colab environment and only run the installs there (also, with a link to open the notebook in colab). E.g. check out [this notebook](https://github.com/skorch-dev/skorch/blob/master/notebooks/Basic_Usage.ipynb). Could you please copy the approach?
2. Right now, you apply the `StandardScaler` directly on the data. How about creating an sklearn `Pipeline` with the `StandardScaler` as the first step and the neural net as the 2nd? That would be a bit cleaner IMO.
3. You can remove the `if __name__ == "__main__":` part.
4. Regarding the visualization, we have a [`simulate` method](https://github.com/skorch-dev/skorch/blob/ad0259b7930ada4f1777240c0a4e0907b57e9387/skorch/callbacks/lr_scheduler.py#L78) in `LRScheduler`. Maybe that can be used instead of having to implement `LRHistoryCallback`.
5. Could you please add an entry to the [notebook/README.md](https://github.com/skorch-dev/skorch/blob/master/notebooks/README.md) and to the [CHANGES.md](https://github.com/skorch-dev/skorch/blob/master/CHANGES.md)?

ParagEkbote commented:
I have modified the notebook, and markdown files as per the suggested changes. Here's a summary:

1. Completed.

2. Completed.

3. Completed.

4. Completed.

5. I have added placeholder links in the notebooks. When this PR gets merged, I will submit a follow-up PR since the notebook needs to be present in the repo before I can add the links to the markdown files.

Could you please review? 

@BenjaminBossan

BenjaminBossan reviewed (changes requested):
Thanks a lot for the updates. I have added some comments, please take a look.

ParagEkbote commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3364 @@
+    "    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=43)\n",
+    "\n",
+    "    # Scaling data using a pipeline\n",
+    "    preprocessing_pipeline = Pipeline([('scaler', StandardScaler())])  \n",
```
I have created the pipeline for making the data preparation process simple and scalable. 

If it an required change, I will modify it.

ParagEkbote replied:
Updated as discussed.

ParagEkbote commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3364 @@
+    "2) ReduceLROnPlateau: Adapts learning rate based on model performance\n",
+    "3) Cosine Annealing: Smoothly reduces learning rate using cosine function\n",
+    "\n",
+    "This notebook uses the simulate method in 'LRScheduler' to visualize the difference between different schedulers.\n",
```
Updated as discussed.

ParagEkbote commented on CHANGES.md:
```diff
@@ -8,6 +8,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
+
+- Added a [notebook]() that shows how to use Learning Rate Scheduler in skorch.(#1074)
```
Updated as discussed.

ParagEkbote commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3364 @@
+    "    scheduler = scheduler_class(optimizer, **optimizer_params)\n",
+    "\n",
+    "    \n",
+    "    criterion = nn.BCEWithLogitsLoss()\n",
```
I have used the BCEwithLogitLoss as it allows to plot the results later in a more proper manner. Previously, I was unable to plot the CosineAnnealing LR with the default loss parameter.

ParagEkbote commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3364 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def simulate_scheduler(scheduler_class, optimizer_params, steps, initial_lr, X_train, y_train, X_val, y_val, input_dim=30, hidden_dim=64):\n",
```
Updated as discussed.

ParagEkbote commented:
Could you please review the changes?

@BenjaminBossan

ParagEkbote commented:
Could you please review the updated changes?

@BenjaminBossan

BenjaminBossan commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3360 @@
+    }
+   ],
+   "source": [
+    "def simulate_scheduler(scheduler_class, optimizer_params, steps, initial_lr):\n",
```
So I checked the skorch code again and I think we can greatly simplify this. Normally, we record the learning rates in the `history`, which means we can just do training and once we finish, we can check `net.history[:, 'event_lr']` to find the learning rates over time. There is no extra code needed to record the learning rates.

The only caveat to this is that we would not record the lr for `ReduceLROnPlateau`. I created a PR to enable this: #1075. So I'd say let's wait for this PR to be merged, then change the notebook to take the learning rates from the history.

ParagEkbote replied:
Please feel free to ping me when #1075, gets merged and I will make the required changes.

BenjaminBossan replied:
The PR is now merged. This means after you finish training a net, you can call `net.history[:, 'event_lr']` to retrieve the learning rates. Please give it a try. Remember to merge the latest master branch of skorch into your branch.

ParagEkbote replied:
Updated as discussed.

BenjaminBossan commented on notebooks/README.md:
```diff
@@ -12,3 +12,4 @@
 * [Hugging Face Vision Transformer](https://nbviewer.org/github/skorch-dev/skorch/blob/master/notebooks/Hugging_Face_VisionTransformer.ipynb)
 * [Skorch Doctor](https://nbviewer.org/github/skorch-dev/skorch/blob/master/notebooks/Skorch_Doctor.ipynb)
 * [Classification with LLMs](https://nbviewer.org/github/skorch-dev/skorch/blob/master/notebooks/LLM_Classifier.ipynb)
+* [Learning Rate Scheduler]()
```
This link can already be prefilled, right?

ParagEkbote replied:
I have updated the missing link.

BenjaminBossan reviewed (changes requested):
Sorry for the delay. I reviewed the PR again and added a few comments. One change is blocked by a new PR, so I'd say let's wait for that PR first. Please take a look.

ParagEkbote commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3360 @@
+    "\n",
+    "\n",
+    "def train_schedulers(X_train, X_test, y_train, y_test, lr=0.05, epochs=1000, hidden_dim=128):\n",
+    "    X_train = torch.tensor(X_train, dtype=torch.float32)\n",
+    "    X_test = torch.tensor(X_test, dtype=torch.float32)\n",
+    "    y_train = torch.tensor(y_train, dtype=torch.float32).view(-1, 1)\n",
+    "    y_test = torch.tensor(y_test, dtype=torch.float32).view(-1, 1)\n",
```
I kept getting the following error.

RuntimeError: mat1 and mat2 must have the same dtype, but got Double and Float

So, I have kept both the tensors in float 32 to avoid the mismatch. If it a necessary change, I will modify it.

ParagEkbote replied:
Updated as discussed.

ParagEkbote commented:
I have addressed the mentioned comments.

Could you please review the updated changes?

cc: @BenjaminBossan

ParagEkbote commented:
Could you please review the updated changes?

cc: @BenjaminBossan

BenjaminBossan commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3350 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a href=\"\" target=\"_blank\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open in Colab\"></a>\n"
```
The colab link is not correct, as the href is missing, right?

ParagEkbote replied:
This a minor nit based off our previous approach. Updated as discussed.

BenjaminBossan commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3350 @@
+    "        print(f\"{scheduler_info['name']} Test Score: {score:.4f}\")\n",
+    "\n",
+    "        # Extract learning rates \n",
+    "        lrs = [event['lr'] for event in net.history if 'lr' in event]\n",
```
This can be simplified to: `lrs = net.history[:, 'event_lr']`.

ParagEkbote replied:
The history attribute does not record the learning rate for all of the schedulers unless set explicitly. I've used the mentioned approach before, and it has not worked out.

BenjaminBossan commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3350 @@
+    "- Reduces the learning rate in a cosine curve over 10 epochs (T_max=10).\n",
+    "- Periodically resets the learning rate, promoting exploration of the loss landscape.\n",
+    "\n",
+    "We will now simulate training the neural network with different LR Schedulers."
```
This is not a "simulation", right? We really do the training. Let's rephrase this.

ParagEkbote replied:
Updated as discussed.

BenjaminBossan commented on notebooks/Learning_Rate_Scheduler.ipynb:
```diff
@@ -0,0 +1,3350 @@
+    "            iterator_train__shuffle=True,\n",
+    "            train_split=None,\n",
+    "            module__input_dim=X_train.shape[1],\n",
+    "            module__hidden_dim=hidden_dim,\n",
```
Personally, I would add `verbose=0` as an argument. That way, we don't have to scroll through multiple screens of the history log and can focus on the essential results. WDYT?

ParagEkbote replied:
I didn't know about this argument, thanks for suggesting!

Updated as discussed.

BenjaminBossan reviewed (changes requested):
Thanks so much for the updates. I went through the notebook again and found only a few small things I'd like to be adjusted. Please check my comments.

ParagEkbote commented:
I have addressed the mentioned comments.

Could you please review the updated changes?

cc: @BenjaminBossan

BenjaminBossan reviewed (approved):
Thanks for making the final changes, the notebook looks great, well done.

ParagEkbote commented:
Thank you for assigning me the issue and being a patient guide.

Could you please see that #1025 gets closed to declutter?

cc: @BenjaminBossan

BenjaminBossan commented:
Thanks for the reminder!
